### ISSUE TITLE
v5develop: Large files and contentscanner

### ISSUE THREAD

**Issue by dbrody2004** [e2guardian/e2guardian#331]
Trying to download a large file (http://releases.ubuntu.com/16.04.3/ubuntu-16.04.3-server-amd64.iso) 
with the clamav contentscanner turned on.  It gets to 20mb and stops.  I've attached the log.  I see this in the file -

hw92: newsize: 131072
hw92: newsize: 262144
hw92: newsize: 524288
hw92: newsize: 1048576
master: busychildren:48 worker Q size:0
hw92: swapping to disk
hw7: written to disk:2097152 total:8388608
hw7: trickle delay - sending X-DGKeepAlive: on

and there is no more hw92 and hw7 lines.  Don't know if swapping to disk is causing the issue.

Also I noticed disabledcontentscanner didn't work but there is a issue filed for that already.  I was able to work around that by commenting out all the contentscanner= lines.

[1.log](https://github.com/e2guardian/e2guardian/files/1590135/1.log)

**Comment by fredbcode** [comment]
> 
> and there is no more hw92 and hw7 lines. Don't know if swapping to
> disk is causing the issue.


Do you have something in syslog ?


> 
> Also I noticed disabledcontentscanner didn't work but there is a
> issue filed for that already. I was able to work around that by
> commenting out all the contentscanner= lines.


Fix pushed


>

**Comment by fredbcode** [comment]
> Trying to download a large file (http://releases.ubuntu.com/16.04.3/ubuntu-16.04.3-server-amd64.iso)

Tested here with 4.1.4 without issue
Do you have something specific from V5 (ICAP ?)

**Comment by dbrody2004** [comment]
Yes this is v5.  This is what I'm observing.  The file appears as a steady incoming stream until 20mB.   The stream reduces to 0mb/s after the initial 20mb.  If you wait long enough, the next 20mB shows up and then again back to 0mb/s and again another wait.  Is the code downloading this next section to disk and then sends it all at once?  I'm trying to understand what is different in the 2nd, 3rd, etc 20mb sections from the first.

**Comment by dbrody2004** [comment]
Figuring this out slowly.  I see the below in the logs.  Correct me if I'm wrong - if the file being downloaded is greater than maxcontentfilecachescansize and a content scanner (i.e. clamav) is specified, e2guardian just gives up downloading the file?  The default for maxcontentfilecachescansize is only 20mb which is not a large file.

Thoughts on fixing this - 
1)  specify a size for maxcontentfilecachescansize that is very large with a limit of disk space available. or 
2)  Other idea?

hw0: swapping to disk
hw0: written to disk:2097152 total:4194304
hw0: written to disk:2097152 total:6291456
hw0: written to disk:2097152 total:8388608
hw0: written to disk:2097152 total:10485760
hw0: written to disk:2097152 total:12582912
hw0: written to disk:2097152 total:14680064
master: busychildren:1 worker Q size:0
hw0: written to disk:2097152 total:16777216
hw0: written to disk:2097152 total:18874368
hw0: written to disk:2097152 total:20971520
hw0: defaultdm: file too big to be scanned, halting download

**Comment by fredbcode** [comment]
Yes there is something interesting here, please can you post here your log file ? 

"file too big to be scanned, halting download" Mean you can't write this object to the ram AND to the disk

A new option like disablecontentscanerror should be useful, disablecontentscanwithlargefile or something like that
I have no time for this next week but, if Philip agree with this idea, I can take a look later.

Note for later: In ConnectionHandler.cpp "::contentFilter" some debug lines must be also in syslog (errors)
Comment from DansGuardian 

// checkme: surely if pausedtoobig is true, we just want to break here?
// the content is larger than max_content_filecache_scan_size if it was downloaded for scanning,
// and larger than max_content_filter_size if not.
// in fact, why don't we check the content length (when it's not -1) before even triggering the download managers?


We should also remove all the useless download manager code -> except default

**Comment by dbrody2004** [comment]
I like the new option.

I think there is something else going on here.  In the default.cpp in downloadmangers, there is a loop that downloads (while ((bytesremaining > 0) || geteverything) { ...} ).  If wantall=true and swappedtodisk=false (i.e under the first default max_content_filecache_scan_size), everything is put into ram.  Eventually when the file is larger to max_content_ramcache_scan_size, the if statement (if (d->buffer_length > o.max_content_ramcache_scan_size) {...}) gets executed changing swappedtodisk=false.  The next time through the loop, wantall=true and swappedtodisk=true causes this statement to be executed (else if (d->tempfilesize > o.max_content_filecache_scan_size) { ...}) which causes the abort.

In other words the logic doesn't allow wantall=true and any file larger than max_content_filecache_scan_size.

**Comment by philipianpearce** [comment]
wantall is an flag internal to the function which indicates that the content length is unknown.  It does not force a complete download.   So the logic appears to be correct.

If the download file is toobig then it will still be virus-checked and then if clean then it will be sent to client and then the rest of the contented tunnelled to client.     So the toobig return from default.cpp does not as such 'abort' the file download.   Maybe the pause is to do with the virus checking of an incomplete file?   

I am also looking at the virus scanning timeouts which do not look quite right.

**Comment by fredbcode** [comment]
Hi Philip,

About "file too big to be scanned, halting download" maybe I'm wrong (I'm out of office right now) but I thought I saw that the download is dropped when e2 can't write this object to the ram and to the disk

hw0: written to disk:2097152 total:20971520
hw0: defaultdm: file too big to be scanned, halting download

In this case it can be useful to add an option to allow the download without scanning, just with a warning in syslog ...

**Comment by dbrody2004** [comment]
Here is a better log of what is happening.  After downloading 20MB (the default in the config file), it stops downloading, sends it the client, and goes back to waiting for connection.

hw0:  in  handle connection
hw0:  waiting connection on http_worker_Q 
hw0:  popped connection from http_worker_Q
hw0:  -got peer connection
hw0: 10.0.19.130
hw0: Start of request header:in
hw0: header:in before getLine - timeout:55000 Line: 1921 Function: in
hw0: firstime: header:in after getLine  Line: 1925 Function: in
hw0: header:size =  6
hw0: first line =  GET /16.04.3/ubuntu-16.04.3-server-amd64.iso HTTP/1.1
hw0: CheckHeader: HTTP/1.1 detected Line: 989 Function: checkheader
hw0: zero contentlength on request due to not POST/PUT 
hw0: CheckHeader flags before normalisation: AP=1 PPC=1 1.1=1 connectionclose=0 CL=0 Line: 1048 Function: checkheader
hw0: CheckHeader flags after normalisation: AP=1 WP=0 Line: 1069 Function: checkheader
hw0: from header url:http://releases.ubuntu.com/16.04.3/ubuntu-16.04.3-server-amd64.iso Line: 1219 Function: getUrl
hw0: setURL: header.front() changed from: GET http://releases.ubuntu.com/16.04.3 Line: 539 Function: setURL4.iso HTTP/1.1
hw0:  to: GET http://releases.ubuntu.com/16.04.3/ubuntu-16.04.3-server-amd64.iso Line: 547 Function: setURL
hw0: isProxyRequest is 0
hw0: Compiling ,*[a-z|A-Z].*
hw0: ...with PCRE 
hw0: from header url:http://releases.ubuntu.com/16.04.3/ubuntu-16.04.3-server-amd64.iso Line: 1219 Function: getUrl
hw0: decoding url Line: 1503 Function: decode
hw0: 3291Start URL http://releases.ubuntu.com/16.04.3/ubuntu-16.04.3-server-amd64.isois_ssl=0ismitm=0
hw0: isProxyRequest is 0 only_ip_auth is 1
hw0:  -Not got persistent credentials for this connection - querying auth plugins
hw0:  -No auth plugins loaded; using defaults & feigning persistency
hw0:  -username: -
hw0:  -filtergroup: 0
hw0:  -About to check for bypass...
hw0: Found cookie: Line: 1293 Function: getCookie
hw0: No bypass cookie Line: 1321 Function: isBypassCookie
hw0:  -Finished bypass checks.
hw0: inURLList: releases.ubuntu.com/16.04.3/ubuntu-16.04.3-server-amd64.iso
hw0: inURLList (processed): releases.ubuntu.com/16.04.3/ubuntu-16.04.3-server-amd64.iso
hw0: inURLList: releases.ubuntu.com/16.04.3/ubuntu-16.04.3-server-amd64.iso
hw0: inURLList (processed): releases.ubuntu.com/16.04.3/ubuntu-16.04.3-server-amd64.iso
hw0: inURLList: releases.ubuntu.com/16.04.3/ubuntu-16.04.3-server-amd64.iso
hw0: inURLList (processed): releases.ubuntu.com/16.04.3/ubuntu-16.04.3-server-amd64.iso
hw0: After StoryB checkrequest isexception 0 gomitm 0 mess_no 0
hw0:   got past line 727 rfo 
hw0: request headerout:GET http://releases.ubuntu.com/16.04.3/ubuntu-16.04.3-server-amd64.iso HTTP/1.1
hw0: request headerout:GET http://releases.ubuntu.com/16.04.3/ubuntu-16.04.3-server-amd64.iso HTTP/1.1
hw0: headertoclient:GET http://releases.ubuntu.com/16.04.3/ubuntu-16.04.3-server-amd64.iso HTTP/1.1
hw0: timeout:55000 Line: 1771 Function: out
hw0: Header written - pstdata_len:0 Line: 1836 Function: out
hw0: Returning from header:out  Line: 1859 Function: out
hw0: Start of response header:in
hw0: header:in before getLine - timeout:61000 Line: 1921 Function: in
hw0: firstime: header:in after getLine  Line: 1925 Function: in
hw0: header:size =  12
hw0: first line =  HTTP/1.1 200 OK
hw0: tp =Content-Length: 865075200
hw0: tp =865075200 Contentlen.int =865075200
hw0: CheckHeader: HTTP/1.1 detected Line: 989 Function: checkheader
hw0: CheckHeader flags before normalisation: AP=1 PPC=1 1.1=1 connectionclose=0 CL=1 Line: 1048 Function: checkheader
hw0: CheckHeader flags after normalisation: AP=1 WP=1 Line: 1069 Function: checkheader
hw0: isProxyRequest is 0
hw0:  -persistPeer: 1
hw0: After StoryB checkresponse0 mess_no 0
hw0:  -Content scanners interested in response data: 1 Line: 1084 Function: handleConnection
hw0: mime type: application/x-iso9660-image Line: 216 Function: isContentType
hw0: mimes result : false Line: 243 Function: isContentType
hw0: mime type: application/x-iso9660-image Line: 216 Function: isContentType
hw0: mimes result : false Line: 243 Function: isContentType
hw0:  -Filtering with expectation of a possible csmessage Line: 2794 Function: check_content
hw0: 
hw0:  -about to get body from proxy
hw0: Got to final download manager so defaulting to always match.
hw0: Inside default download manager plugin  icap=0
hw0: tranencodeing is 
hw0: bytes remaining is 865075200
hw0: blocksize: 32768
hw0: newsize: 32768
hw0: newsize: 32768
hw0: newsize: 65536
hw0: newsize: 131072
hw0: newsize: 262144
hw0: newsize: 524288
hw0: newsize: 1048576
hw0: swapping to disk
hw0: written to disk:2097152 total:4194304
hw0: sending first line of header first
hw0: response headerout:HTTP/1.1 200 OK
hw0: response headerout:HTTP/1.1 200 OK
hw0: headertoclient:HTTP/1.1 200 OK
hw0: timeout:61000 Line: 1771 Function: out
hw0: trickle delay - sending X-DGKeepAlive: on
hw0: written to disk:2097152 total:6291456
hw0: trickle delay - sending X-DGKeepAlive: on
hw0: written to disk:2097152 total:8388608
hw0: trickle delay - sending X-DGKeepAlive: on
hw0: written to disk:2097152 total:10485760
hw0: trickle delay - sending X-DGKeepAlive: on
hw0: written to disk:2097152 total:12582912
hw0: trickle delay - sending X-DGKeepAlive: on
hw0: written to disk:2097152 total:14680064
hw0: trickle delay - sending X-DGKeepAlive: on
hw0: written to disk:2097152 total:16777216
hw0: trickle delay - sending X-DGKeepAlive: on
hw0: written to disk:2097152 total:18874368
hw0: trickle delay - sending X-DGKeepAlive: on
hw0: written to disk:2097152 total:20971520
hw0: trickle delay - sending X-DGKeepAlive: on
hw0: defaultdm: file too big to be scanned, halting download
hw0: Leaving default download manager plugin
hw0:  -got PARTIAL body 
hw0:  -content length large so skipping content scanning (virus) filtering
hw0:  -Skipping content filtering: hw0:  -Content too largehw0: 
hw0:  -Skipping content modification: hw0:  -Content too largehw0: 
hw0:  Returning from content checking
hw0: End content check isitNaughty is  0
hw0: response headerout:HTTP/1.1 200 OK
hw0: response headerout:HTTP/1.1 200 OK
hw0: headertoclient:HTTP/1.1 200 OK
hw0: timeout:61000 Line: 1771 Function: out
hw0: Header written - pstdata_len:0 Line: 1836 Function: out
hw0: Returning from header:out  Line: 1859 Function: out
hw0: Sending 20971520 bytes from temp file (0 already sent)
hw0: reading temp file rc:2097152
hw0: total sent from temp:2097152
hw0: reading temp file rc:2097152
hw0: total sent from temp:4194304
hw0: reading temp file rc:2097152
hw0: total sent from temp:6291456
hw0: reading temp file rc:2097152
hw0: total sent from temp:8388608
hw0: reading temp file rc:2097152
hw0: total sent from temp:10485760
hw0: reading temp file rc:2097152
hw0: total sent from temp:12582912
hw0: reading temp file rc:2097152
hw0: total sent from temp:14680064
hw0: reading temp file rc:2097152
hw0: total sent from temp:16777216
hw0: reading temp file rc:2097152
hw0: total sent from temp:18874368
hw0: reading temp file rc:2097152
hw0: total sent from temp:20971520
hw0:  -Forwarding body to client : Upfailure is 0
hw0: Compiling ,*[a-z|A-Z].*
hw0: ...with PCRE 
hw0:  -persisting (count 1)
hw0:  - 10.0.19.130
hw0: Start of request header:in
hw0: header:in before getLine - timeout:55000 Line: 1921 Function: in
hw0: firstime: header:in after getLine  Line: 1925 Function: in
hw0: firstime: header:in after getLine: rc: 0 truncated: 1 Line: 1930 Function: in
hw0:  -Persistent connection closed
hw0:  -Attempting graceful connection close Line: 1171 Function: handleConnection
hw0: handle_peer returned: 0
hw0:  in  handle connection
hw0:  waiting connection on http_worker_Q

**Comment by philipianpearce** [comment]
Thanks for the debug trace - I will walk through the code using it to see where the logic is wrong.

**Comment by philipianpearce** [comment]
Problem was that DataBuffer::out() return was undefined - was returning false when should have returned true.   Please test in case of any other issues.

**Comment by dbrody2004** [comment]
Nice catch - works now.  Thanks
